### ISSUE TITLE
Sharing: Facebook counts differ for http and https URLs

### ISSUE THREAD

**Issue by wjkowalski** [Automattic/jetpack#2536]
Before changing to the internal API for Facebook share counts, we would request both the HTTP and HTTPS counts from Facebook. After the revert back to using Facebook’s API, we only count the current URL. For sites that operate on either protocol or have switched, this results in a decent drop in their output. For example, all WP.com sites forced HTTPS recently, effectively resetting their Facebook counts.

**Comment by kraftbj** [comment]
cc: @aduth

**Comment by aduth** [comment]
Related: #2495

Reasons we shouldn't sum counts:

- More often than not, Facebook will report the same count for both protocols, resulting in a falsely doubled count (see #1560)
- We should avoid inconsistencies between the numbers reported by custom-style sharing buttons and official sharing buttons

Website owners should:

- Ensure that their website uses a single canonical URL. Serving both HTTPS and HTTP for a single domain is not a good practice -- in the same way that you should not serve both WWW and top-level domain, and instead redirect to one or the other.
- Submit feedback to Facebook that 301 redirects should be considered in share counts reported

One option we could pursue is to add a filter enabling developers to override the URL used in requesting share counts. Currently, this logic is almost entirely client-side JavaScript and is not friendly for developers to override. Noting this, I won't have the bandwidth to take a deeper look at this for at least a couple months.

**Comment by kraftbj** [comment]
Reading through the old tickets and testing:
1. It seems like if the item was shared/liked when a HTTP->HTTPS 301 is in place, Facebook automatically sums. e.g. Site A is https://example.com/post-slug/ and I share http://example.com/post-slug/ -- this will bump FB's share count for the https site. I assume FB respects the canonical and "assigns" the action to the canonical URL.

2. For sites that were canonically changed, e.g. WordPress.com changing all *.wordpress.com sites from http to https, Facebook does not pull both the http and https values—it just returns the canonical's value, which based on 1, has already been credited with http shares after the canonical was changed to https. 

In other words, it seems that Facebook credits the canonical URL at the time of the action with the action. I suppose for Jetpack's users, that's one thing since that was their option to do so. Hurts more for WP.com users who didn't have an option (unless using a custom domain).

Filter(s) would be nice. Thinking aloud, that would let a developer write up code, knowing when the site changed canonical style, could dynamically determine when to sum the canonical URL with some other URL(s). In my case, I could write a filter if the post date is before May 1, sum my canonical with my previous URL structure, etc.

We should be able to backtrack—roughly at least—when wp.com was switched over and implement something similar.

It would be different behavior than the official buttons, but I'd argue this is better behavior, as long as we have it documented. :)

**Comment by aduth** [comment]
>In other words, it seems that Facebook credits the canonical URL at the time of the action with the action. 

Correct, and apologies that my previous message was unclear. Facebook __does__ follow redirects when a share is made, but it will not retroactively consider share counts for redirects put in place after the initial share action.

**Comment by kraftbj** [comment]
Noting too that it seems like if the share was made with a HTTPS canonical, FB credits both.

Compare:
http://graph.facebook.com/?id=https://kraft.im/2015/06/fathers-day-truth/
http://graph.facebook.com/?id=http://kraft.im/2015/06/fathers-day-truth/
(both report 35 shares)
That URL has always been HTTPS with a HTTP->HTTPS 301.

http://graph.facebook.com/?id=http://kraft.wordpress.com/2014/04/02/put-down-the-phone/
http://graph.facebook.com/?id=https://kraft.wordpress.com/2014/04/02/put-down-the-phone/
(http reports 25, https report none)
The post was published when the canonical was HTTP. No likes/shares since the URL changed to https.

**Comment by kraftbj** [comment]
Letting this soak, but not a quick fix. Moving the milestone.

**Comment by timdream** [comment]
> The post was published when the canonical was HTTP. No likes/shares since the URL changed to https.

Does Automattic has contacts to Facebook that could ask them to expose some kind of API allowing people to move their canonical URLs, at least from HTTP to HTTPS?

**Comment by kraftbj** [comment]
@aduth can weigh in on the current situation. I believe we explored various options with them, but been awhile now.

**Comment by aduth** [comment]
We've reached out to Facebook contacts multiple times in the past regarding this issue, and have tried to make a strong case for transitioning counts for redirects (or at least treating HTTP/HTTPS as the same), especially as the web continues to move toward HTTPS everywhere. They had seemed receptive to the idea, though we got the sense that it was not an active priority for any of their teams at the time. I'll see to following-up to determine whether any progress has been made or if we can push for it to become a focus.

**Comment by timdream** [comment]
@aduth Thank you very much for the update.

**Comment by aduth** [comment]
Oh, I had also meant to share a link to Facebook's documentation where they describe workarounds for preserving counts when a URL changes:

https://developers.facebook.com/docs/sharing/webmasters/crawler#updating

**Comment by jeherve** [comment]
Also reported here:
https://wordpress.org/support/topic/sharing-social-share-counts-are-gone/

**Comment by jeherve** [comment]
Also reported here:
https://wordpress.org/support/topic/lost-no-of-shares-on-my-blog-posts-on-moving-from-http-to-https/#post-8735767

**Comment by georgian29** [comment]
I had the same problem and I've found a fix here: https://cognitiveseo.com/blog/13431/recover-facebook-shares-https/

**Comment by kraftbj** [comment]
Thanks for sharing that @georgian29 !

Since this issue is, all in all, in Facebook's court to change globally and there's not a good way for us to do it on a massive scale, I'm going to close this issue. Please reopen if circumstances change.

**Comment by htdat** [comment]
New tix: 699175-zen

**Comment by kristarella** [comment]
> Oh, I had also meant to share a link to Facebook's documentation where they describe workarounds for preserving counts when a URL changes:

We are now doing this on atomic sites (https://github.com/Automattic/atomic/issues/377#issuecomment-350874220).

I wrote a plugin that will force the `og:url` to be HTTP and I'm pretty sure it fixes the issue for posts that were previously shared using HTTP:
https://gist.github.com/kristarella/e0efbac9011708f3c9ee34106b6f202a

**Comment by kristarella** [comment]
I have a problem with a subset of posts that were shared using HTTPS and now that atomic sites are being redirected to HTTP for the Facebook crawler, this subset don't have the right share count.


- https://cgfewston.me/2017/08/03/you-are-the-universe-discovering-your-cosmic-self-and-why-it-matters-2017-by-deepak-chopra-and-menas-kafatos-where-thoughts-come-from/
- https://cgfewston.me/2017/08/10/do-not-say-we-have-nothing-2016-by-madeleine-thien-the-book-of-records/
- https://cgfewston.me/2017/10/02/the-sense-of-style-2014-by-steven-pinker-the-importance-of-precise-language/
- https://cgfewston.me/2017/10/08/1q84-2009-by-haruki-murakami-the-two-moons/
- https://cgfewston.me/2017/10/23/his-bloody-project-2015-by-graeme-macrae-burnet-questions-of-innocence/
- https://cgfewston.me/2017/11/28/the-stories-of-john-cheever-1978-by-john-cheever-the-creation-of-fiction-as-art/

If you view them in the [object debugger](https://developers.facebook.com/tools/debug/og/object/) rather than the sharing debugger you can see the previous share count at the bottom of the page.

I think we need to contact Facebook about this, because we are already implementing their suggestions for dealing with this issue and they are still broken after many rescrapes.

810401-zen
